monty-31/src/x86_64_avx512/utils.rs

```diff
@@ -0,0 +1,221 @@
+
+        // As hi < r2^{j - N} < P, the output is always in [0, P]. It is equal to P only when input t = 0.
+        x86_64::_mm512_sub_epi32(lo_shft, hi)
+    }
+}
+
+/// Multiply a vector of Monty31 field elements in canonical form by -2**{-8}.
+/// # Safety
+///
+/// The prime P must be of the form P = r * 2^j + 1 with r odd and r < 2^7.
+/// Input must be given in canonical form.
+/// Output is not in canonical form, outputs are only guaranteed to lie in (-P, P).
+#[inline(always)]
+pub unsafe fn mul_neg_2_exp_neg_8_avx512<
+    TAD: TwoAdicData + PackedMontyParameters,
+    const N_PRIME: u32,
+>(
+    input: __m512i,
+) -> __m512i {
+    /*
+        We want this to compile to:
+            mov         lo_shft,            P           This can be a mov or a load. It will not affect throughput/latency.
+            vpsrld      hi,                 val,        8
+            vptestmd    lo_MASK,            val,        2^8 - 1
+            vpmaddubsw  lo_x_r,             val,        [r; 16]
+            vpslld      lo_shft{lo_MASK},   lo_x_r,     j - 8
+            vpsubd      res,                lo_shft,    hi
+        throughput: 3
+        latency: 7
+    */
+    unsafe {
+        assert_eq!(8 + N_PRIME, TAD::TWO_ADICITY as u32); // Compiler removes this provided it is satisfied.
+
+        let odd_factor: __m512i = transmute([TAD::ODD_FACTOR; 16]); // This is [r; 16]. Compiler realises this is a constant.
+        let hi = x86_64::_mm512_srli_epi32::<8>(input);
+
+        let mask: __m512i = transmute([(1 << 8) - 1; 16]); // Compiler realises this is a constant.
+
+        // Determine the non 0 values of lo.
+        let lo_mask = x86_64::_mm512_test_epi32_mask(input, mask);
+
+        // Whilst it generically does something else, provided
+        // each entry of odd_factor is < 2^7, _mm512_maddubs_epi16
+        // performs an element wise multiplication of odd_factor with
+        // the bottom 8 bits of input interpreted as an unsigned integer
+        // Thus lo_x_r contains lo * r.
+        let lo_x_r = x86_64::_mm512_maddubs_epi16(input, odd_factor);
+
+        // When lo = 0, lo_shft = P
+        // When lo > 0, lo_shft = r2^{j - N} x_lo
+        let lo_shft = x86_64::_mm512_mask_slli_epi32::<N_PRIME>(TAD::PACKED_P, lo_mask, lo_x_r);
+
+        // As hi < r2^{j - N} < P, the output is always in [0, P]. It is equal to P only when input t = 0.
+        x86_64::_mm512_sub_epi32(lo_shft, hi)
+    }
+}
+
+/// Multiply a vector of Monty31 field elements in canonical form by -2**{-N} where P = 2^31 - 2^N + 1.
+/// # Safety
+///
+/// The prime P must have the form P = 2^31 - 2^N + 1.
+/// Input must be given in canonical form.
+/// Output is not in canonical form, outputs are only guaranteed to lie in (-P, P).
+#[inline(always)]
+pub unsafe fn mul_neg_2_exp_neg_two_adicity_avx512<
+    TAD: TwoAdicData + PackedMontyParameters,
+    const N: u32,
+    const N_PRIME: u32,
```

Comment:
I definitely agree with you but unfortunately this isn't allowed by rust. I get the error:

Generic parameters may not be used in const operations.

We could add the nightly feature generic_const_exprs which would fix that but my understanding was we wanted to avoid that feature for the time being.

Reply:
Ah yep, I forgot once again how dumb Rust can be some times. :roll_eyes: Carry on.